### DISCUSSION TITLE
Need inputs on mocking "useDefaultApiClient" in jest test cases for React

### DISCUSSION THREAD

**Discussion by himan085** [aws/aws-pdk#716] [General]
Hi, the generated code by PDK provides a method `useDefaultApiClient` for making authenticated calls using SigV4. I need some inputs on understanding how to mock this method for my test cases using Jest in React.
There are 2 versions of mocking that I have experimented with:

**Version 1**:
```
const mockCreateApi = jest.fn();

jest.mock("../../../../hooks/use-default-api-client", () => ({
  useDefaultApiClient: () => {
    return {
      create: mockCreateApi,
    };
  },
}));

it("successfully renders", () => {
  mockCreateApi.mockResolvedValue("success"); // resolving value inside test case
  ...
});
```
---

**Version 2**:
```
const mockClient = {
  create: () => {
    return Promise.resolve("success");
  },
};

jest.mock("../../../../hooks/use-default-api-client", () => ({
  useDefaultApiClient: () => mockClient,
}));

it("successfully renders", () => {
  // test cases always resolve to "success" as specified by the mockClient
  ...
});
```

I'm mostly using `version 1` across my code base. For one my specific source codes, the jest test cases are getting stuck with no termination. I tried to wait around 10mins but the tests were simply stuck, not even timing out. However when I changed the mock to use `version 2` the test cases worked as expected. The source code in the browser is also working as expected.

Note: I verified there aren't any infinite loops by logging in both test cases and in the browser console.

I need some inputs on what might be going wrong here ? Any help is appreciated, thanks!
Let me know if more inputs or the source code is also required.

**Comment by cogwirrel** [comment]
Hi Himanshu, thanks for reaching out!

I'm afraid I'm not sure why you'd be experiencing tests hanging with `version 1` and not `version 2` - it's hard to tell without the full codebase (and I am certainly not a jest expert!). Both of the methods look ok to me!

One possibility is that the hook is being invoked outside of your test somehow, so you need to provide a mock resolved value to begin with? You could try adding `mockCreateApi.mockResolvedValue("success");` to just after where `mockCreateApi` is declared?

Sorry to not be of more help!

Thanks,
Jack
